package frc.robot.subsystems.hood;

import static org.wpilib.units.Units.Amps;
import static org.wpilib.units.Units.Degrees;
import static org.wpilib.units.Units.Radians;
import static org.wpilib.units.Units.RadiansPerSecond;
import static org.wpilib.units.Units.RotationsPerSecond;
import static org.wpilib.units.Units.RotationsPerSecondPerSecond;
import static org.wpilib.units.Units.Seconds;
import static org.wpilib.units.Units.Volts;

import coppercore.controls.state_machine.StateMachine;
import coppercore.math.Lazy;
import coppercore.monitors.TotalCurrentCalculator;
import coppercore.parameter_tools.LoggedTunableNumber;
import coppercore.wpilib_interface.MonitorWithAlert.MonitorWithAlertBuilder;
import coppercore.wpilib_interface.MonitoredSubsystem;
import coppercore.wpilib_interface.UnitUtils;
import coppercore.wpilib_interface.subsystems.motors.MotorIO;
import coppercore.wpilib_interface.subsystems.motors.MotorInputsAutoLogged;
import coppercore.wpilib_interface.subsystems.motors.profile.MotionProfileConfig;
import coppercore.wpilib_interface.tuning.TestModeManager;
import org.wpilib.units.AngularVelocityUnit;
import org.wpilib.units.measure.Angle;
import org.wpilib.units.measure.AngularVelocity;
import org.wpilib.driverstation.Alert;
import org.wpilib.driverstation.internal.DriverStationBackend;
import org.wpilib.system.RobotController;
import frc.robot.CoordinationLayer.ShotMode;
import frc.robot.DependencyOrderedExecutor;
import frc.robot.DependencyOrderedExecutor.ActionKey;
import frc.robot.constants.JsonConstants;
import frc.robot.subsystems.hood.HoodState.HomingWaitForButtonState;
import frc.robot.subsystems.hood.HoodState.HomingWaitForMovementState;
import frc.robot.subsystems.hood.HoodState.HomingWaitForStoppingState;
import frc.robot.subsystems.hood.HoodState.IdleState;
import frc.robot.subsystems.hood.HoodState.TargetAngleState;
import frc.robot.subsystems.hood.HoodState.TargetExitPitchState;
import frc.robot.subsystems.hood.HoodState.TestModeState;
import frc.robot.util.StateMachineDump;
import org.littletonrobotics.junction.AutoLogOutput;
import org.littletonrobotics.junction.AutoLogOutputManager;
import org.littletonrobotics.junction.Logger;

/**
 * The HoodSubsystem class defines the Hood subsystem, which controls the hardware for the hood on
 * the aimer of our shooter superstructure. It uses simple closed loop control to accomplish its
 * tasks.
 */
public class HoodSubsystem extends MonitoredSubsystem {
  private enum HoodAction {
    /** Coasts the hood and waits for input */
    Idle,
    /** Targets a certain pitch (exit angle), commanded by the supervisor layer */
    TargetExitPitch,
    /** Targets a certain angle, commanded by the supervisor layer */
    TargetAngle,
  }

  private final MotorIO motor;
  private final MotorInputsAutoLogged inputs = new MotorInputsAutoLogged();

  public static final ActionKey UPDATE_INPUTS = new ActionKey("HoodSubsystem::updateInputs");

  /**
   * Whether or not the homing switch is currently pressed. This value is updated by the
   * CoordinationLayer via the setIsHomingSwitchPressed method.
   */
  private boolean isHomingSwitchPressed = false;

  // State machine and states
  private final StateMachine<HoodSubsystem> stateMachine;

  private final HoodState homingWaitForButtonState;
  private final HoodState homingWaitForMovementState;
  private final HoodState homingWaitForStoppingState;
  private final HoodState idleState;
  private final HoodState targetExitPitchState;
  private final HoodState targetAngleState;
  private final HoodState testModeState;

  // Test mode
  TestModeManager<TestMode> testModeManager = new TestModeManager<TestMode>("Hood", TestMode.class);

  // Tunable numbers
  private final Lazy<LoggedTunableNumber> hoodKP;
  private final Lazy<LoggedTunableNumber> hoodKI;
  private final Lazy<LoggedTunableNumber> hoodKD;

  private final Lazy<LoggedTunableNumber> hoodKS;
  private final Lazy<LoggedTunableNumber> hoodKV;
  private final Lazy<LoggedTunableNumber> hoodKA;
  private final Lazy<LoggedTunableNumber> hoodKG;

  private final Lazy<LoggedTunableNumber> hoodExpoKV;
  private final Lazy<LoggedTunableNumber> hoodExpoKA;

  private final Lazy<LoggedTunableNumber> hoodTuningSetpointDegrees;
  private final Lazy<LoggedTunableNumber> hoodTuningAmps;
  private final Lazy<LoggedTunableNumber> hoodTuningVolts;

  // State variables
  /**
   * The last action requested of the Hood. This is the action that the hood should target, but only
   * if the state machine allows. For example, it should only track a pitch after homing.
   */
  @AutoLogOutput(key = "Hood/action")
  private HoodAction requestedAction = HoodAction.Idle;

  private Angle goalExitPitch =
      JsonConstants.hoodConstants
          .minHoodAngle
          .plus(JsonConstants.hoodConstants.mechanismAngleToExitAngle);

  private Angle goalAngle = JsonConstants.hoodConstants.minHoodAngle;

  // Dependencies (values from other subsystems/coordination layer passed in by the coordination
  // layer via setters)
  /**
   * Whether or not shooting is enabled. Set by the CoordinationLayer using setShootingEnabled, and
   * used to determine whether the hood should stow.
   */
  private boolean shootingEnabled = false;

  /**
   * Whether or not the hood should currently stow for the trench. Set by the CoordinationLayer
   * using setShouldStowForTrench
   */
  private boolean shouldStowForTrench = false;

  /**
   * Whether or not the hood should currently stow to protect the intake. Set by the
   * CoordinationLayer using setShouldStowForIntake.
   */
  private boolean shouldStowForIntakeOrDefense = false;

  public HoodSubsystem(DependencyOrderedExecutor dependencyOrderedExecutor, MotorIO motor) {
    this.motor = motor;

    motor.setRequestUpdateFrequency(JsonConstants.hoodConstants.hoodRequestUpdateFrequency);

    addMonitor(
        new MonitorWithAlertBuilder()
            .withName("HoodMotorDisconnected")
            .withAlertText("Hood motor disconnected")
            .withAlertType(Alert.Level.HIGH)
            .withTimeToFault(JsonConstants.hoodConstants.disconnectedDebounceTimeSeconds)
            .withLoggingEnabled(true)
            .withStickyness(false)
            .withIsStateValidSupplier(() -> inputs.connected)
            .withFaultCallback(() -> {})
            .build());

    // Define state machine and transitions
    stateMachine = new StateMachine<>(this);

    homingWaitForButtonState = stateMachine.registerState(new HomingWaitForButtonState());
    homingWaitForMovementState = stateMachine.registerState(new HomingWaitForMovementState());
    homingWaitForStoppingState = stateMachine.registerState(new HomingWaitForStoppingState());
    idleState = stateMachine.registerState(new IdleState());
    targetExitPitchState = stateMachine.registerState(new TargetExitPitchState());
    targetAngleState = stateMachine.registerState(new TargetAngleState());
    testModeState = stateMachine.registerState(new TestModeState());

    homingWaitForButtonState
        .when(hood -> hood.isHomingSwitchPressed(), "Homing switch is pressed")
        .transitionTo(idleState);
    homingWaitForButtonState
        .when(() -> DriverStationBackend.isEnabled(), "Robot is enabled")
        .transitionTo(homingWaitForMovementState);

    homingWaitForMovementState
        .when((hood) -> hood.isMoving(), "Is moving")
        .transitionTo(homingWaitForStoppingState);
    homingWaitForMovementState
        .whenTimeout(JsonConstants.hoodConstants.homingMaxUnmovingTime)
        .transitionTo(homingWaitForStoppingState);

    homingWaitForStoppingState
        .when(hood -> !hood.isMoving(), "Is not moving")
        .transitionTo(idleState);

    idleState.when(hood -> hood.isHoodTestMode(), "Is hood test mode").transitionTo(testModeState);
    idleState
        .when(hood -> hood.requestedAction == HoodAction.TargetExitPitch, "Action == TargetPitch")
        .transitionTo(targetExitPitchState);
    idleState
        .when(hood -> hood.requestedAction == HoodAction.TargetAngle, "Action == TargetAngle")
        .transitionTo(targetAngleState);

    targetExitPitchState
        .when(hood -> hood.requestedAction != HoodAction.TargetExitPitch, "Action != TargetPitch")
        .transitionTo(idleState);

    targetAngleState
        .when(hood -> hood.requestedAction != HoodAction.TargetAngle, "Action != TargetAngle")
        .transitionTo(idleState);

    testModeState
        .when(hood -> !hood.isHoodTestMode(), "Isn't hood test mode")
        .transitionTo(idleState);

    stateMachine.setState(homingWaitForButtonState);
    StateMachineDump.write("hood", stateMachine);

    // Initialize tunable numbers for test modes
    hoodKP =
        new Lazy<>(
            () ->
                new LoggedTunableNumber("HoodTunables/HoodKP", JsonConstants.hoodConstants.hoodKP));
    hoodKI =
        new Lazy<>(
            () ->
                new LoggedTunableNumber("HoodTunables/HoodKI", JsonConstants.hoodConstants.hoodKI));
    hoodKD =
        new Lazy<>(
            () ->
                new LoggedTunableNumber("HoodTunables/HoodKD", JsonConstants.hoodConstants.hoodKD));

    hoodKS =
        new Lazy<>(
            () ->
                new LoggedTunableNumber("HoodTunables/HoodKS", JsonConstants.hoodConstants.hoodKS));
    hoodKG =
        new Lazy<>(
            () ->
                new LoggedTunableNumber("HoodTunables/HoodKG", JsonConstants.hoodConstants.hoodKG));
    hoodKV =
        new Lazy<>(
            () ->
                new LoggedTunableNumber("HoodTunables/HoodKV", JsonConstants.hoodConstants.hoodKV));
    hoodKA =
        new Lazy<>(
            () ->
                new LoggedTunableNumber("HoodTunables/HoodKA", JsonConstants.hoodConstants.hoodKA));

    hoodExpoKV =
        new Lazy<>(
            () ->
                new LoggedTunableNumber(
                    "HoodTunables/HoodExpoKV", JsonConstants.hoodConstants.hoodExpoKV));
    hoodExpoKA =
        new Lazy<>(
            () ->
                new LoggedTunableNumber(
                    "HoodTunables/HoodExpoKA", JsonConstants.hoodConstants.hoodExpoKA));

    hoodTuningSetpointDegrees =
        new Lazy<>(
            () ->
                new LoggedTunableNumber(
                    "HoodTunables/HoodTuningSetpointDegrees",
                    JsonConstants.hoodConstants.minHoodAngle.in(Degrees)));
    hoodTuningAmps = new Lazy<>(() -> new LoggedTunableNumber("HoodTunables/HoodAmps", 0.0));
    hoodTuningVolts = new Lazy<>(() -> new LoggedTunableNumber("HoodTunables/HoodVolts", 0.0));

    dependencyOrderedExecutor.registerAction(UPDATE_INPUTS, this::updateInputs);

    AutoLogOutputManager.addObject(this);
  }

  private void updateInputs() {
    motor.updateInputs(inputs);
    Logger.processInputs("Hood/inputs", inputs);

    TotalCurrentCalculator.recordCurrent(hashCode(), inputs.supplyCurrentAmps);

    // Log values with units so that AdvantageScope can understand them correctly
    Logger.recordOutput("Hood/closedLoopReferenceRadians", inputs.closedLoopReference);
    Logger.recordOutput("Hood/closedLoopReferenceSlopeRadPerSec", inputs.closedLoopReferenceSlope);

    // For some reason, AutoLogOutput doesn't log the unit correctly, so we have to log it here.
    Logger.recordOutput("Hood/exitAngleRadians", getCurrentExitPitch().in(Radians));

    Logger.recordOutput(
        "Hood/bottomAngleRadians",
        inputs.positionRadians - JsonConstants.hoodConstants.minHoodAngle.in(Radians));
  }

  @Override
  public void monitoredPeriodic() {
    long startTimeUs = RobotController.getTime();

    Logger.recordOutput("Hood/state", stateMachine.getCurrentState().getName());
    stateMachine.periodic();

    long endTimeUs = RobotController.getTime();
    if (JsonConstants.featureFlags.logPeriodicTiming) {
      Logger.recordOutput("PeriodicTime/hoodMs", (endTimeUs - startTimeUs) / 1000.0);
    }
  }

  /**
   * Polls for test-mode specific actions (such as updating gains from network tables)
   *
   * <p>This method must be called by the test mode state, as it does not run automatically
   */
  protected void testPeriodic() {
    switch (testModeManager.getTestMode()) {
      case HoodClosedLoopTuning -> {
        LoggedTunableNumber.ifChanged(
            hashCode(),
            (pid_sgva) -> {
              JsonConstants.hoodConstants.hoodKP = pid_sgva[0];
              JsonConstants.hoodConstants.hoodKI = pid_sgva[1];
              JsonConstants.hoodConstants.hoodKD = pid_sgva[2];
              JsonConstants.hoodConstants.hoodKS = pid_sgva[3];
              JsonConstants.hoodConstants.hoodKG = pid_sgva[4];
              JsonConstants.hoodConstants.hoodKV = pid_sgva[5];
              JsonConstants.hoodConstants.hoodKA = pid_sgva[6];

              motor.setGains(
                  JsonConstants.hoodConstants.hoodKP,
                  JsonConstants.hoodConstants.hoodKI,
                  JsonConstants.hoodConstants.hoodKD,
                  JsonConstants.hoodConstants.hoodKS,
                  JsonConstants.hoodConstants.hoodKG,
                  JsonConstants.hoodConstants.hoodKV,
                  JsonConstants.hoodConstants.hoodKA);
            },
            hoodKP.get(),
            hoodKI.get(),
            hoodKD.get(),
            hoodKS.get(),
            hoodKG.get(),
            hoodKV.get(),
            hoodKA.get());

        LoggedTunableNumber.ifChanged(
            hashCode(),
            (expoConstraintsVA) -> {
              motor.setProfileConstraints(
                  MotionProfileConfig.immutable(
                      RotationsPerSecond.zero(),
                      RotationsPerSecondPerSecond.zero(),
                      RotationsPerSecondPerSecond.zero().div(Seconds.of(1.0)),
                      Volts.of(expoConstraintsVA[0]).div(RotationsPerSecond.of(1)),
                      Volts.of(expoConstraintsVA[1]).div(RotationsPerSecondPerSecond.of(1))));

              JsonConstants.hoodConstants.hoodExpoKV = expoConstraintsVA[0];
              JsonConstants.hoodConstants.hoodExpoKA = expoConstraintsVA[1];
            },
            hoodExpoKV.get(),
            hoodExpoKA.get());

        // Make sure that the mechanism doesn't break itself by controlling outside of its safe
        // range of motion, even in test mode
        clampAndControlToAngle(Degrees.of(hoodTuningSetpointDegrees.get().getAsDouble()));
      }
      case HoodCurrentTuning -> {
        motor.controlOpenLoopCurrent(Amps.of(hoodTuningAmps.get().getAsDouble()));
      }
      case HoodVoltageTuning -> {
        motor.controlOpenLoopVoltage(Volts.of(hoodTuningVolts.get().getAsDouble()));
      }
      default -> {}
    }
  }

  /**
   * Returns `true` if the absolute value of the hood's velocity is greater than or equal to the
   * hood homing movement threshold, `false` otherwise
   *
   * @return `true` if the absolute value of the hood's velocity is greater than or equal to the
   *     hood homing movement threshold, `false` otherwise
   */
  protected boolean isMoving() {
    final AngularVelocityUnit velocityComparisonUnit = RadiansPerSecond;

    return getVelocity().abs(velocityComparisonUnit)
        >= JsonConstants.turretConstants.homingMovementThreshold.in(velocityComparisonUnit);
  }

  /**
   * Sets the hood's current position to the homed position. Should be called whenever homing states
   * determine that the system is at its homed position.
   */
  protected void onHomePositionReached() {
    motor.setCurrentPosition(JsonConstants.hoodConstants.minHoodAngle);
  }

  /**
   * Updates the hood subsystem on whether the homing switch is pressed. This should only be called
   * by a coordinator/supervisor-layer action scheduled with the DependencyOrderedExecutor.
   *
   * @param isHomingSwitchPressed True if the homing switch is pressed (hood should assume it has
   *     homed), false if the switch isn't pressed.
   */
  public void setIsHomingSwitchPressed(boolean isHomingSwitchPressed) {
    this.isHomingSwitchPressed = isHomingSwitchPressed;
  }

  /**
   * Updates the hood subsystem on whether shooting is enabled. This should only be called by the
   * coordination layer, and is used to determine whether the hood should stow or not (if shooting,
   * the hood will aim, and if not, it will stow).
   *
   * @param shootingEnabled {@code true} if shooting is enabled and the hood should aim, {@code
   *     false} if shooting is disabled
   */
  public void setShootingEnabled(boolean shootingEnabled) {
    this.shootingEnabled = shootingEnabled;
  }

  /**
   * Updates the hood subsystem on whether it should stow to go under the trench. This should only
   * be called by the coordination layer.
   *
   * @param shouldStowForTrench {@code true} if the hood should stow to go under the trench, {@code
   *     false} otherwise.
   */
  public void setShouldStowForTrench(boolean shouldStowForTrench) {
    Logger.recordOutput("Hood/shouldStowForTrench", shouldStowForTrench);
    this.shouldStowForTrench = shouldStowForTrench;
  }

  /**
   * Updates the hood subsystem on whether it should stow to prepare for the intake being stowed.
   * This should only be called by the coordination layer.
   *
   * <p>This should also be set to true during defense mode, as it will stow the hood and then keep
   * it applying a neutral request against the hardstop, which will save some power.
   *
   * @param shouldStowForIntakeOrDefense {@code true} if the intake is enabled and above the hood
   *     stow threshold, {@code false} otherwise
   */
  public void setShouldStowForIntakeOrDefense(boolean shouldStowForIntakeOrDefense) {
    Logger.recordOutput("Hood/shouldStowForIntakeOrDefense", shouldStowForIntakeOrDefense);
    this.shouldStowForIntakeOrDefense = shouldStowForIntakeOrDefense;
  }

  /**
   * Returns whether or not the homing switch is currently pressed (or was pressed when its inputs
   * were last read from hardware.)
   *
   * @return True if the homing switch is pressed (the mechanism should consider itself homed),
   *     false if not
   */
  protected boolean isHomingSwitchPressed() {
    return isHomingSwitchPressed;
  }

  /**
   * Apply the homing voltage defined in HoodConstants to gently home the hood into its bottom
   * hardstop
   */
  protected void applyHomingVoltage() {
    motor.controlOpenLoopVoltage(JsonConstants.hoodConstants.homingVoltage);
  }

  /**
   * Get the mechanism velocity of the hood
   *
   * @return An AngularVelocity representing the velocity of the physical hood
   */
  public AngularVelocity getVelocity() {
    return RadiansPerSecond.of(inputs.velocityRadiansPerSecond);
  }

  /** Applies a CoastOut/neutral request. */
  protected void coast() {
    motor.controlCoast();
  }

  protected void controlToGoalExitPitch() {
    Logger.recordOutput("Hood/goalPitchRadians", goalExitPitch.in(Radians));
    Angle goalAngle =
        Degrees.of(90)
            .minus(goalExitPitch)
            .minus(JsonConstants.hoodConstants.mechanismAngleToExitAngle);
    Logger.recordOutput("Hood/goalAngleRadians", goalAngle.in(Radians));
    clampAndControlToAngle(goalAngle);
  }

  protected void controlToGoalAngle() {
    Logger.recordOutput("Hood/goalAngleRadians", goalAngle.in(Radians));
    clampAndControlToAngle(goalAngle);
  }

  /**
   * Given a target angle, clamp it to be within the allowed bounds and then control the mechanism
   * toward it
   *
   * <p>This includes automatically clamping max angle/stowing the hood when necessary.
   *
   * @param goalAngle The Angle to target
   */
  private void clampAndControlToAngle(Angle goalAngle) {
    boolean shouldStowForShootingDisabled = !shootingEnabled && !DriverStationBackend.isUtility();
    boolean shouldStow =
        shouldStowForShootingDisabled || shouldStowForTrench || shouldStowForIntakeOrDefense;

    // If we need to stow, clamp angle to always be set to minHoodAngle.
    Angle maxAngle =
        shouldStow
            ? JsonConstants.hoodConstants.minHoodAngle
            : JsonConstants.hoodConstants.maxHoodAngle;
    Angle clampedGoalAngle =
        UnitUtils.clampMeasure(goalAngle, JsonConstants.hoodConstants.minHoodAngle, maxAngle);
    Logger.recordOutput("Hood/clampedGoalAngleRadians", clampedGoalAngle.in(Radians));

    if (shouldStow
        && Math.abs(inputs.positionRadians - JsonConstants.hoodConstants.minHoodAngle.in(Radians))
            < JsonConstants.hoodConstants.hoodStowEpsilon.in(Radians)) {
      // Brake mode to save power when stowed once the hood is actually stowed
      motor.controlBrake();
    } else {
      // Use unprofiled position control because:
      // 1. We don't really risk breaking the mechanism by going too fast (aluminum gears)
      // 2. The motion profile isn't actually active during shooting because closed loop reference
      // doesn't react to the error caused by fuel going through, so we need good unprofiled PID
      // gains
      // anyway.
      // 3. When testing this, it actually performed better and overshot less with no profile than
      // with a profile.
      motor.controlToPositionUnprofiled(clampedGoalAngle);
    }
  }

  /**
   * Get the current position of the hood (NOT EXIT ANGLE) in radians
   *
   * @return An Angle containing the current angle of the hood (in terms of center of mass).
   */
  public Angle getCurrentAngle() {
    return Radians.of(inputs.positionRadians);
  }

  /**
   * Get the current fuel exit pitch/exit angle based on the position of the hood
   *
   * @return An Angle representing the current exit angle of a fuel being shot from the hood.
   */
  public Angle getCurrentExitPitch() {
    return Degrees.of(90)
        .minus(getCurrentAngle().plus(JsonConstants.hoodConstants.mechanismAngleToExitAngle));
  }

  public boolean isHoodTestMode() {
    return testModeManager.isInTestMode();
  }

  /**
   * Returns whether or not the hood is currently aimed at its goal angle
   *
   * @param shotMode A ShotMode to determine if we are shooting at the hub (tight thresholds) or
   *     passing (loose thresholds)
   * @return {@code true} if the hood is targeting an angle or pitch and it's at that goal, {@code
   *     false} otherwise.
   */
  public boolean isAimedCorrectly(ShotMode shotMode) {
    Angle threshold =
        switch (shotMode) {
          case Hub -> JsonConstants.hoodConstants.hoodSetpointEpsilon;
          case Pass -> JsonConstants.hoodConstants.hoodPassingSetpointEpsilon;
        };

    boolean aimedCorrectly =
        switch (requestedAction) {
          case TargetAngle -> getCurrentAngle().isNear(goalAngle, threshold);
          case TargetExitPitch -> getCurrentExitPitch().isNear(goalExitPitch, threshold);
          case Idle -> false;
        };

    Logger.recordOutput("Hood/isAimedCorrectly", aimedCorrectly);
    return aimedCorrectly;
  }

  /**
   * Sets the goal exit pitch/exit angle of the hood. Note that this value is NOT the goal angle of
   * the hood, but rather the desired fuel exit angle while shooting.
   *
   * <p>This method updates the hood's current action, so that as soon as homing is completed, it
   * will target the pitch requested. This means that it can safely be called at any time,
   * regardless of homing status.
   *
   * @param goalPitch An Angle containing the desired angle above the horizon (zero being horizontal
   *     and 90 degrees being a vertical shot) at which a fuel should exit the hood.
   */
  public void targetExitPitch(Angle goalPitch) {
    this.requestedAction = HoodAction.TargetExitPitch;
    this.goalExitPitch = goalPitch;
  }

  /**
   * Sets the goal angle of the hood. Note that this is a hood angle, NOT an exit angle for the
   * projectile.
   *
   * <p>This method updates the hood's current action, so that as soon as homing is completed, it
   * will target the angle requested. This means that it can safely be called at any time,
   * regardless of homing status.
   *
   * @param angleRadians A double containing the desired hood angle in radians.
   */
  public void targetAngleRadians(double angleRadians) {
    this.requestedAction = HoodAction.TargetAngle;
    this.goalAngle = new Angle(angleRadians, 1.0, Radians);
  }
}
